presto-spark-classloader-interface/src/main/java/com/facebook/presto/spark/classloader_interface/PrestoSparkZipRdd.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.spark.classloader_interface;
+
+import org.apache.spark.Partition;
+import org.apache.spark.SparkContext;
+import org.apache.spark.TaskContext;
+import org.apache.spark.rdd.RDD;
+import org.apache.spark.rdd.ZippedPartitionsBaseRDD;
+import org.apache.spark.rdd.ZippedPartitionsPartition;
+import scala.Tuple2;
+import scala.collection.JavaConversions;
+import scala.collection.Seq;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.function.Function;
+import java.util.stream.IntStream;
+
+import static java.util.Collections.unmodifiableList;
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.toList;
+import static scala.collection.JavaConversions.asScalaBuffer;
+import static scala.collection.JavaConversions.asScalaIterator;
+import static scala.collection.JavaConversions.seqAsJavaList;
+
+public class PrestoSparkZipRdd
+        extends ZippedPartitionsBaseRDD<Tuple2<Integer, PrestoSparkRow>>
```

Comment:
`ZippedPartitionsBaseRDD` seems to be a `private[static]` class : https://github.com/apache/spark/blob/fd2d55c9919ece5463377bc6f45f2cdb8bf90515/core/src/main/scala/org/apache/spark/rdd/ZippedPartitionsRDD.scala#L45  Shall we extend it?

cc @sameeragarwal

Reply:
Ideally - no. This is a private API. However public API does not allow zipping of more than 4 RDD's. That's why we are making this tradeoff.

I acknowledge that we might looks forward compatibility at some point due to using this private API. However unfortunately I don't see how can we zip more than 4 partition in a way other than that.

This is tested with Spark 2.0 and Spark 2.4.3 and it works. If at some point we discover that compatibility is broken for some version of Spark we will try to figure out how to fix it.

Meanwhile we can also try to contribute N-way ZIP partition to open source Spark.

Reply:
@arhimondr : 
> Meanwhile we can also try to contribute N-way ZIP partition to open source Spark.

That could be the long-term solution. Another worthwhile investigation next half could be to understand how does SparkSQL implement it without N-way ZIP partition. I also see SparkSQL implements some specialized SparkSQL RDD (unfortunately, in Scala). Thus they don't have this Java-Scala interaction design problem :)

Reply:
They implemented by calling zip partitions (or join / etc:) multiple times, as they dont really need to have 1 task per stage, since the exchange is row based anyway.

Reply:
@arhimondr : Even SparkSQL is already row-based, calling `zipPartition` might also cause unnecessary materialization (depends on implementation). I also feel it could limit the efficiency of whole stage codegen without having all the zipped partitions all in once...